packages/zoe/src/contracts/operaConcertTicket.js

```diff
@@ -0,0 +1,174 @@
+  - Smart Contract: 
+    - mints the tickets
+    - provides the seats
+  - Auditorium (unique contract seat, usually taken by the contract creator): the person hosting 
+  the Opera show, selling the tickets and getting the payment back
+  - Ticket buyers (contract seat created on demand): 
+    - can see the available opera show seats
+    - can consult the terms
+    - can redeem the zoe invite with the proper payment to get the ticket back
+
+  ERTP and Zoe are considered to be the most highly trusted pieces of code by everyone
+  They are more trusted than the code of this contract
+  As a consequence, they are going to be leveraged as much as possible by this contract
+  to increase its trustworthiness and by the contract users
+  
+*/
+
+export const makeContract = harden(zoe => {
+  // Create the internal ticket mint
+  const { issuer, mint, amountMath: ticketAmountMath } = produceIssuer(
+    'Opera tickets',
+    'set',
+  );
+
+  const {
+    terms: { show, start, count },
+    issuerKeywordRecord: { Money: moneyIssuer },
+  } = zoe.getInstanceRecord();
+
+  const moneyAmountMath = moneyIssuer.getAmountMath();
+
+  return zoe.addNewIssuer(issuer, 'Ticket').then(() => {
+    // create Zoe helpers after zoe.addNewIssuer because of https://github.com/Agoric/agoric-sdk/issues/802
+    const { rejectOffer } = makeZoeHelpers(zoe);
+
+    // Mint tickets inside the contract
+    // In a more realistic contract, the Auditorium would certainly mint the tickets themselves
+    // but because of a current technical limitation when running the Agoric stack on a blockchain,
+    // minting has to happen inside a Zoe contract https://github.com/Agoric/agoric-sdk/issues/821
+
+    // Mint the contract ahead-of-time (instead of on-demand)
+    // This way, they can be passed to Zoe + ERTP who will be doing the bookkeeping
+    // of which tickets have been sold and which tickets are still for sale
+    const ticketsAmount = ticketAmountMath.make(
+      harden(
+        Array(count)
+          .fill()
+          .map((_, i) => {
+            const ticketNumber = i + 1;
+            return harden({
+              show,
+              start,
+              number: ticketNumber,
+            });
+          }),
+      ),
+    );
+    const ticketsPayment = mint.mintPayment(ticketsAmount);
+
+    const {
+      invite: contractSelfInvite,
+      inviteHandle: contractOfferHandle,
+    } = zoe.makeInvite();
+    // the contract creates an offer {give: tickets, want: nothing} with the tickets
+    return zoe
+      .getZoeService()
+      .redeem(
+        contractSelfInvite,
+        harden({ give: { Ticket: ticketsAmount } }),
+        harden({ Ticket: ticketsPayment }),
+      )
+      .then(() => {
+        const auditoriumSeat = harden({
+          // this is meant to be called right after redeem
+          // eventually, it might be done automatically: https://github.com/Agoric/agoric-sdk/issues/717
+          afterRedeem() {
```

Comment:
I suspect we need this method, but can you walk me through why? It's because we cannot reallocate until the invite for the auditorium has been redeemed, right?

Reply:
No, i think it's not needed actually. The code of this method can be moved to right after 
```js
const {
  invite: auditoriumInvite,
  inviteHandle: auditoriumOfferHandle,
} = zoe.makeInvite(auditoriumSeat);
```

because it doesn't need anything more than `auditoriumOfferHandle`

Reply:
I'll move the code and remove the method

Reply:
oh... I'm discovering that the method is needed
and for a reason that resonates with the discussion about [how people get `offerHandle`s](https://github.com/Agoric/agoric-sdk/issues/810#issuecomment-608460808)

Currently in `afterRedeem`, i get the current allocation for the `auditoriumOfferHandle`. And that's not possible to do that before the auditorium has redeemed the invite because calls to `zoe.getOffer(handle)` only works after `redeem`. It's only when redeem happens that the offerHandle is used for something useful:

https://github.com/Agoric/agoric-sdk/blob/e2c311b4d536f18d6966b0fc10721b581fd1368a/packages/zoe/src/zoe.js#L448-L458

(line 456 which uses the handle as key of a table entry that is consulted in `zoe.getOffer`)

I think this is further argument to create separate inviteHandle and offerHandle, and use each for its own purpose (inviteHandle for redeem+seat and offerHandle for interaction with allocations)

Reply:
I'm going to keep the afterRedeem method for now

Reply:
Ok, sounds good, and thanks for the investigation.